src/include/sof/coherent.h

```diff
@@ -20,11 +20,33 @@
 #define __coherent __attribute__((packed, aligned(DCACHE_LINE_SIZE)))
 
 /*
+ * The coherent API allows optimized access to memory by multiple cores, using
+ * cache, taking care about coherence. The intended use is to let cores acquire
+ * ownership of such shared objects, use them, and then release them, possibly
+ * to be re-acquired by other cores. Such shared objects must only be accessed
+ * via this API. It can only be used with dynamically allocated objects, not
+ * with objects from .data or .bss sections.
+ *
+ * The API assumes, that in the beginning no core has cache lines associated
+ * with that memory. That is true for dynamically allocated memory, because when
+ * such memory is freed, its cache is invalidated. The first call must be
+ * coherent_init(), which initializes the header. If the object will be used by
```

Comment:
We assume that first use for DATA and BSS is **the** first use.

Reply:
@lgirdwood well, maybe a more precise wording would be "it is impossible to enforce" this API for `.data` / `.bss`, because anybody can grab pointers to those objects at any time and use them. So, you only can rely on the program flow to guarantee the correct use. E.g. if some core begins to use it before it has been initialised or similar.

Reply:
Whereas with heap data you're safe: when you allocate memory, no core has caches for it, because when it was freed last time, all caches have been dropped (as long as that previous allocation wasn't sharing memory without using the coherent API...). Your code is running on core X, which allocates memory and presumably immediately initialises it for the coherent API. From this point it's safe for use by any core, using the API. Until the last user releases it and all caches are dropped and then once more during`rfree()`

Reply:
@lgirdwood yes, agree. If it must be done, it can be done, but one has to be very careful with this...